internal/scheduler/scheduling/optimiser/node_scheduler.go

```diff
@@ -0,0 +1,285 @@
+package optimiser
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/armadaproject/armada/internal/common/logging"
+	"github.com/armadaproject/armada/internal/common/util"
+	"github.com/armadaproject/armada/internal/scheduler/internaltypes"
+	"github.com/armadaproject/armada/internal/scheduler/jobdb"
+	"github.com/armadaproject/armada/internal/scheduler/nodedb"
+	"github.com/armadaproject/armada/internal/scheduler/scheduling/context"
+	"github.com/armadaproject/armada/internal/server/configuration"
+)
+
+type PreemptingNodeScheduler struct {
+	jobDb                   jobdb.JobRepository
+	maximumJobSizeToPreempt *internaltypes.ResourceList
+}
+
+type NodeScheduler interface {
+	Schedule(schedContext *SchedulingContext, jctx *context.JobSchedulingContext, node *internaltypes.Node) (*nodeSchedulingResult, error)
+}
+
+func NewPreemptingNodeScheduler(jobDb jobdb.JobRepository, maximumJobSizeToPreempt *internaltypes.ResourceList) *PreemptingNodeScheduler {
+	return &PreemptingNodeScheduler{
+		jobDb:                   jobDb,
+		maximumJobSizeToPreempt: maximumJobSizeToPreempt,
+	}
+}
+
+// Schedule
+// This function is responsible for determining if a job can be scheduled on a node, and the cost to do so
+// High level steps:
+// - Determine all the jobs that can be preempted
+// - Order these jobs in the ideal order to preempt them in
+//   - Group these jobs by queue and order them by the order that queue would want those jobs ordered
+//   - This will give you N ordered lists, one of each queue
+//   - Combine these ordered lists into a "global" order, which is the ideal order to preempt jobs in
+//
+// - Attempt to schedule the new job on the node, preempting one job at a time until the job fit
+func (n *PreemptingNodeScheduler) Schedule(schedContext *SchedulingContext, jctx *context.JobSchedulingContext, node *internaltypes.Node) (*nodeSchedulingResult, error) {
+	met, _, err := nodedb.StaticJobRequirementsMet(node, jctx)
+	if err != nil {
+		return nil, err
+	}
+	if !met {
+		return &nodeSchedulingResult{
+			jctx:      jctx,
+			node:      node,
+			resultId:  util.NewULID(),
+			scheduled: false,
+		}, nil
+	}
+
+	availableResource := node.AllocatableByPriority[internaltypes.EvictedPriority]
+	if !jctx.Job.AllResourceRequirements().Exceeds(availableResource) {
```

Comment:
Should use `job.KubernetesResourceRequirements()` as floating resources will never be available on the node.